core/src/main/java/jenkins/model/GlobalBuildDiscarderStrategy.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2019 Daniel Beck
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.model;
+
+import hudson.Extension;
+import hudson.model.Job;
+import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+
+/**
+ * Apply a user-specified build discarder periodically on all jobs.
+ */
+public class GlobalBuildDiscarderStrategy extends BackgroundBuildDiscarderStrategy {
+
+    private BuildDiscarder discarder;
```

Comment:
Public class, all methods public, class neither `final` nor  marked `@NoExternalUse`, _but_... the key instance variable is `private`? 

I know there's a culture of `always private` around a lot of the Jenkins codebase, but this seems like an odd one out among the rest of an otherwise-extensible & reusable class.

Reply:
Good point. It wasn't written with subclasses in mind; there's little benefit to doing that anyway, given the ratio of necessary new code (`Descriptor`!) to reusable code. I'm fairly sure I've experienced hierarchies in Jenkins actively hostile to subclasses due to how the generics worked out, plus there's always backward compatibility concerns because of plugins written a decade ago, and the problem of missing UI support for new supertype features… best not to start here if there's a perfectly good `ExtensionPoint` to work with. So I would not expect that this is something you'd extend in a plugin; rather, start with the abstract class directly. That's the most stable interface for extension in plugins, and no changes to this type (a future `[x] Skip jobs with a configured build discarder` perhaps?) could affect your subtype. There's just a handful of additional lines of duplication.

Reply:
> It wasn't written with subclasses in mind;

IMO, all non-`final` classes are implicitly written with subclasses in mind, because it's Java. The actively-hostile-to-subclasses parts of Jenkins feel so alien compared to the rest of my Java experience where extension is ... part of how one builds in the language.

Perhaps at least add a JavaDoc comment to the (abridged) effect of your above comment here "Don't extend this to make your own strageies; extend the abstract `BackgroundBuildDiscarderStrategy`"

Reply:
Probably just going to make it `@Restricted` to prevent any use from plugins; cannot think of anything that would be expected besides configuration-as-code, and that ignores class-level `@Restricted`.